File: apps/downloadx/src/throttle.ts
/**
 * Token-bucket bandwidth limiter.
 *
 * Usage: each chunk calls `consume(bytes)` right before it writes the data it
 * just read. If the bucket has enough tokens, resolves immediately; otherwise
 * resolves after the shortage has refilled.
 *
 * A single instance is shared by all chunks of a download, so the cap applies
 * to the aggregate bandwidth rather than per-chunk.
 *
 * `capacityBytesPerSec === 0` disables throttling — `consume` becomes a no-op.
 */
export class Throttle {
  private tokens: number;
  private lastRefillAt: number;
  private readonly queue: Array<{
    need: number;
    resolve: () => void;
    reject: (err: Error) => void;
    signal?: AbortSignal;
    onAbort?: () => void;
  }> = [];
  private disposed = false;

  constructor(
    private capacity: number,
    private readonly now: () => number = Date.now,
    private readonly schedule: (ms: number, cb: () => void) => void = (ms, cb) => {
      setTimeout(cb, ms);
    },
  ) {
    if (capacity < 0) throw new Error(`Throttle: capacity must be >= 0 (got ${capacity})`);
    this.tokens = capacity;
    this.lastRefillAt = now();
  }

  /** Change the cap live (e.g. user toggles speedLimit mid-download). */
  setCapacity(capacity: number): void {
    if (capacity < 0) throw new Error(`Throttle: capacity must be >= 0 (got ${capacity})`);
    this.refill();
    const previous = this.capacity;
    this.capacity = capacity;
    if (capacity === 0) {
      // Switching to unlimited — release everything in-flight immediately.
      this.tokens = 0;
      for (const entry of this.queue.splice(0)) {
        entry.signal?.removeEventListener('abort', entry.onAbort!);
        entry.resolve();
      }
      return;
    }
    if (capacity > previous) {
      // On raise, top the bucket up to the new ceiling so queued waiters can
      // make progress right away instead of having to wait for a drip refill.
      this.tokens = capacity;
    } else if (this.tokens > capacity) {
      // Shrink reservoir to fit the new cap.
      this.tokens = capacity;
    }
    this.drainQueue();
  }

  get capacityBytesPerSec(): number {
    return this.capacity;
  }

  /**
   * Wait until `bytes` tokens are available, then deduct them. Resolves
   * immediately when `capacity === 0` (unlimited).
   *
   * If `bytes > capacity`, the request is satisfied in a single shot after
   * waiting `bytes / capacity` seconds — we don't partition the request so
   * the caller can treat consume() as atomic.
   */
  consume(bytes: number, signal?: AbortSignal): Promise<void> {
    if (bytes <= 0) return Promise.resolve();
    if (this.capacity === 0) return Promise.resolve();
    if (this.disposed) return Promise.reject(new Error('Throttle disposed'));

    this.refill();
    if (this.tokens >= bytes) {
      this.tokens -= bytes;
      return Promise.resolve();
    }

    return new Promise<void>((resolve, reject) => {
      if (signal?.aborted) {
        reject(toAbortError(signal.reason));
        return;
      }
      const entry: (typeof this.queue)[number] = { need: bytes, resolve, reject };
      if (signal !== undefined) {
        entry.signal = signal;
        entry.onAbort = (): void => {
          const idx = this.queue.indexOf(entry);
          if (idx !== -1) this.queue.splice(idx, 1);
          reject(toAbortError(signal.reason));
        };
        signal.addEventListener('abort', entry.onAbort, { once: true });
      }
      this.queue.push(entry);
      this.scheduleDrain();
    });
  }

  dispose(): void {
    this.disposed = true;
    for (const entry of this.queue) {
      entry.signal?.removeEventListener('abort', entry.onAbort!);
      entry.reject(new Error('Throttle disposed'));
    }
    this.queue.length = 0;
  }

  private refill(): void {
    if (this.capacity === 0) return;
    const now = this.now();
    const elapsed = now - this.lastRefillAt;
    if (elapsed <= 0) return;
    this.tokens = Math.min(this.capacity, this.tokens + (elapsed * this.capacity) / 1000);
    this.lastRefillAt = now;
  }

  private drainQueue(): void {
    this.refill();
    while (this.queue.length > 0) {
      const head = this.queue[0];
      if (head === undefined) break;
      if (this.tokens < head.need) break;
      this.tokens -= head.need;
      this.queue.shift();
      head.signal?.removeEventListener('abort', head.onAbort!);
      head.resolve();
    }
    if (this.queue.length > 0) this.scheduleDrain();
  }

  private scheduleDrain(): void {
    const head = this.queue[0];
    if (head === undefined || this.capacity === 0) return;
    const missing = head.need - this.tokens;
    if (missing <= 0) {
      this.drainQueue();
      return;
    }
    const waitMs = Math.max(5, Math.ceil((missing * 1000) / this.capacity));
    this.schedule(waitMs, () => this.drainQueue());
  }
}

function toAbortError(reason: unknown): Error {
  if (reason instanceof Error) return reason;
  const err = new Error('Aborted');
  err.name = 'AbortError';
  return err;
}

File: apps/downloadx_dart/lib/src/throttle.dart
import 'dart:async';

import 'io.dart';

/// Token-bucket bandwidth limiter.
///
/// Each chunk calls [consume] right before writing the data it just read. If
/// the bucket has enough tokens, it resolves immediately; otherwise it resolves
/// once the shortage has refilled. One instance is shared by all chunks of a
/// download, so the cap applies to aggregate bandwidth.
///
/// `capacity == 0` disables throttling — [consume] becomes a no-op.
class Throttle {
  num _capacity;
  double _tokens;
  int _lastRefillAt;
  bool _disposed = false;

  final int Function() _now;
  final void Function(int ms, void Function() cb) _schedule;
  final List<_Waiter> _queue = [];

  Throttle(
    num capacity, {
    int Function()? now,
    void Function(int ms, void Function() cb)? schedule,
  })  : _capacity = capacity,
        _now = now ?? _defaultNow,
        _schedule = schedule ?? _defaultSchedule,
        _tokens = capacity.toDouble(),
        _lastRefillAt = (now ?? _defaultNow)() {
    if (capacity < 0) {
      throw ArgumentError('Throttle: capacity must be >= 0 (got $capacity)');
    }
  }

  /// Change the cap live (e.g. user toggles speedLimit mid-download).
  void setCapacity(num capacity) {
    if (capacity < 0) {
      throw ArgumentError('Throttle: capacity must be >= 0 (got $capacity)');
    }
    _refill();
    final previous = _capacity;
    _capacity = capacity;
    if (capacity == 0) {
      // Unlimited — release everything in-flight immediately.
      _tokens = 0;
      final pending = List<_Waiter>.of(_queue);
      _queue.clear();
      for (final entry in pending) {
        entry.dispose?.call();
        entry.resolve();
      }
      return;
    }
    if (capacity > previous) {
      // On raise, top the bucket up so queued waiters make progress at once.
      _tokens = capacity.toDouble();
    } else if (_tokens > capacity) {
      _tokens = capacity.toDouble();
    }
    _drainQueue();
  }

  num get capacityBytesPerSec => _capacity;

  /// Wait until [bytes] tokens are available, then deduct them. Resolves
  /// immediately when capacity is 0 (unlimited).
  Future<void> consume(int bytes, [CancelToken? signal]) {
    if (bytes <= 0) return Future.value();
    if (_capacity == 0) return Future.value();
    if (_disposed) return Future.error(StateError('Throttle disposed'));

    _refill();
    if (_tokens >= bytes) {
      _tokens -= bytes;
      return Future.value();
    }

    final completer = Completer<void>();
    if (signal?.isCancelled ?? false) {
      return Future.error(signal!.reason);
    }
    final waiter = _Waiter(bytes, completer);
    if (signal != null) {
      waiter.dispose = signal.onCancel(() {
        _queue.remove(waiter);
        if (!completer.isCompleted) completer.completeError(signal.reason);
      });
    }
    _queue.add(waiter);
    _scheduleDrain();
    return completer.future;
  }

  void dispose() {
    _disposed = true;
    for (final entry in _queue) {
      entry.dispose?.call();
      if (!entry.completer.isCompleted) {
        entry.completer.completeError(StateError('Throttle disposed'));
      }
    }
    _queue.clear();
  }

  void _refill() {
    if (_capacity == 0) return;
    final now = _now();
    final elapsed = now - _lastRefillAt;
    if (elapsed <= 0) return;
    final next = _tokens + (elapsed * _capacity) / 1000;
    _tokens = next < _capacity ? next : _capacity.toDouble();
    _lastRefillAt = now;
  }

  void _drainQueue() {
    _refill();
    while (_queue.isNotEmpty) {
      final head = _queue.first;
      if (_tokens < head.need) break;
      _tokens -= head.need;
      _queue.removeAt(0);
      head.dispose?.call();
      head.resolve();
    }
    if (_queue.isNotEmpty) _scheduleDrain();
  }

  void _scheduleDrain() {
    if (_queue.isEmpty || _capacity == 0) return;
    final head = _queue.first;
    final missing = head.need - _tokens;
    if (missing <= 0) {
      _drainQueue();
      return;
    }
    final waitMs = (missing * 1000 / _capacity).ceil();
    _schedule(waitMs < 5 ? 5 : waitMs, _drainQueue);
  }
}

class _Waiter {
  final int need;
  final Completer<void> completer;
  void Function()? dispose;
  _Waiter(this.need, this.completer);
  void resolve() {
    if (!completer.isCompleted) completer.complete();
  }
}

int _defaultNow() => DateTime.now().millisecondsSinceEpoch;
void _defaultSchedule(int ms, void Function() cb) =>
    Timer(Duration(milliseconds: ms), cb);
